### PR TITLE
Fix relation constraints and cleaning

### PR DESCRIPTION
**Pull request by cedric-anne**
| Q             | A
| ------------- | ---
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | -

I made 3 commits.
1. Reindent and alphabetically reorder existing mapping.
2. Fixes
3. Reindent fixed mapping.

Review should be made on 2nd commit.

Changes: 
- Add "_" prefix and add missing recursive relation for `CommonTreeDropdown` tables
- Add "_" prefix and add missing relations handled in-code (by `$item::cleanDBonPurge()` method or using `$forward_entity_to` property value)
- Factorize deletion of childs and relations on item deletion
- Fix missing cleanup when deleting items:
  - costs related to deleted change,
  - changes related to deleted certificate, dc room, enclosure pdu, rack or software license,
  - certificate items related to deleted computer, network equipement, peripheral, printer, software license or user,
  - rss feed relation with deleted entity, group, profile or user,
  - knowbase item translation related to deleted knowbase item,
  - notification template translation related to deleted notification template,
  - change items related to deleted pdu, rack, software license,
  - locks on objects related to deleted user,
  - costs related to deleted problem,
  - costs related to deleted project,
  - queued notifications related to deleted notification template,
  - saved search alerts related to deleted save search,
- Fix missing entity forward to information:
  - from project to project cost,
- Fix relations mapping to enable detection/transfer when deleting item and enable check on ability to disable recursivity of an item:
  - between API client and entity,
  - between business criticities and entity,
  - between change cost and budget,
  - between budget and budget type,
  - between certificate and certificate type,
  - between certificate type and entity,
  - between change task and editor user, group and task template,
  - between change validation and user,
  - between computer antivirus and manufacturer,
  - between contract cost and budget,
  - between device and model, type, entity, location and state,
  - between document item and user,
  - between domain and entity,
  - between infocom and business criticity,
  - between generic device and generic device item,
  - between motherboard device and motherboard device item,
  - between simcard device and line,
  - between device item and location and state,
  - between operation system and achitecture, edition and kernel version,
  - between itil solution and solution type, users and ticket followup,
  - between knowbase sub items and users,
  - between line operator and entity,
  - between line and entity, user, group, line operator, location, state and line type,
  - between ethernet network port and network card device,
  - between fiber channel network and network car device, network port and net point,
  - between network migrations and entity, net point and network interface,
  - between network port and network interface,
  - between wifi network port and network card device,
  - between notepad and users,
  - between notification template and notification template translation,
  - between refused emails and mail collector,
  - between ola level and ola,
  - between ola and calendar,
  - between operation system kernel version and operating system kernel,
  - between planning recall and user,
  - between problem stask adn editor user, group and task template,
  - between profile and ticket template,
  - between rss feed profile and entity,
  - between project cost and project, budget and entity,
  - between project task and project task template,
  - between project task template and project task, project state and project task type,
  - between queued notification and entity,
  - between rss feed and user,
  - between sla and calendar,
  - between software license type and entity,
  - between state and entity,
  - between task category and knowbase item category,
  - between task template and user and group,
  - between ticket cost and budget,
  - between ticket followup and editor user,
  - between recurrent ticket and calendar,
  - between user and auth methods,
  - between ticket task and task template, group and editor user,
  - between ticket validation and users,
- Remove obsolete relations

### CODE REVIEW

File: inc/cartridgeitem.class.php [hunk 1]
```diff
@@ -80,11 +80,12 @@ function getPostAdditionalInfosForName() {
...
-      $class->cleanDBonItemDelete($this->getType(), $this->fields['id']);
+      $this->deleteChildrenAndRelationsFromDb(
+         [
+            Cartridge::class,
+            CartridgeItem_PrinterModel::class,
+         ]
+      );
 
       $class = new Alert();
       $class->cleanDBonItemDelete($this->getType(), $this->fields['id']);
```
Comment (trasher) [inline inc/cartridgeitem.class.php]: Why this one has not been replaced?

File: inc/commondbtm.class.php [hunk 2]
```diff
@@ -862,6 +862,28 @@ function cleanDBonPurge() {
...
+    * @param array $relations_classes List of classname on which deletion will be done
+    *                                 Classes needs to extends CommonDBConnexity.
+    *
+    * @return void
+    **/
+   protected function deleteChildrenAndRelationsFromDb(array $relations_classes) {
+
+      foreach ($relations_classes as $classname) {
+         if (!is_a($classname, CommonDBConnexity::class, true)) {
+            continue;
```
Comment (trasher) [inline inc/commondbtm.class.php]: This would lend in silently failing cases... Log at least a warning.

File: inc/commonitilobject.class.php [hunk 3]
```diff
@@ -584,23 +584,23 @@ function countActiveObjectsForSupplier($suppliers_id) {
 
    function cleanDBonPurge() {
 
-      if (!empty($this->grouplinkclass)) {
-         $class = new $this->grouplinkclass();
-         $class->cleanDBonItemDelete($this->getType(), $this->fields['id']);
+      $link_classes = [
+         ITILSolution::class
```
Comment (trasher) [inline inc/commonitilobject.class.php]: Maybe tasks and documents should be handled here. And ITILFollowup (soon in master)

File: inc/cartridgeitem.class.php (hunk 1 above)
Comment (cedric-anne) [reply inc/cartridgeitem.class.php]: Presence of the `cleanDBonItemDelete` method is a naming coincidence as `Alert` class does not extends CommonDBConnexity.
Maybe it should, but I do not really know all impacts of using `CommonDBChild` as base class for `Alert`.

Comment (trasher) [reply inc/cartridgeitem.class.php]: Ok, I've seen quite similar cases in the PR that were an evidence. OK for me, do not change.

File: inc/commonitilobject.class.php (hunk 3 above)
Comment (trasher) [reply inc/commonitilobject.class.php]: Documents relations are handled in CommonDBTM; and it's not possible to handle tasks "proprerly" here (needs some refactoring before).

File: inc/commondbtm.class.php (hunk 2 above)
Comment (cedric-anne) [reply inc/commondbtm.class.php]: Done.

**Comment by cedric-anne** [comment]
I added tests to validate that all tables and fields are existing.

**Comment by trasher** [comment]
There are conflicts on `inc/relation.constant.php` trying to cherry-pick the commit on master branch. I'm not able to solve them.

**Comment by cedric-anne** [comment]
Done in #4773 .

**Pull request merged**